### DISCUSSION TITLE
Difficulty parsing a text file with complex grammar

### DISCUSSION THREAD

**Discussion by patrobinson** [alecthomas/participle#252] [Q&A] [answered]
I am trying to parse some poker hands, which utilise very complex grammer it seems.
Here's an example hand

```
Poker Hand #OM58920214: PLO-5 ($0.01/$0.02) - 2022/07/14 10:58:16
Table 'PLO-5White35' 6-max Seat #6 is the button
Seat 1: acd0710 ($2.34 in chips)
Seat 2: edbdf39d ($2.14 in chips)
Seat 3: 181cc43b ($2.38 in chips)
Seat 4: Hero ($2 in chips)
Seat 5: deae4305 ($2 in chips)
Seat 6: fe5c8313 ($2.62 in chips)
acd0710: posts small blind $0.01
edbdf39d: posts big blind $0.02
*** HOLE CARDS ***
Dealt to acd0710 
Dealt to edbdf39d 
Dealt to 181cc43b 
Dealt to Hero [Kc Js 4d 9c 5h]
Dealt to deae4305 
Dealt to fe5c8313 
181cc43b: raises $0.05 to $0.07
Hero: folds
deae4305: folds
fe5c8313: folds
acd0710: folds
edbdf39d: folds
Uncalled bet ($0.05) returned to 181cc43b
*** SHOWDOWN ***
181cc43b collected $0.05 from pot
*** SUMMARY ***
Total pot $0.05 | Rake $0 | Jackpot $0 | Bingo $0
Seat 1: acd0710 (small blind) folded before Flop
Seat 2: edbdf39d (big blind) folded before Flop
Seat 3: 181cc43b collected ($0.05)
Seat 4: Hero folded before Flop (didn't bet)
Seat 5: deae4305 folded before Flop (didn't bet)
Seat 6: fe5c8313 (button) folded before Flop (didn't bet)
```

I've managed to parse the first two lines, but I'm stuck on the third. My progress is very slow, as I find it difficult to comprehend the parser's internal state and why it's failing. I'm using a custom Lexer and regex to match individual fields.

My code is available [here](https://github.com/patrobinson/punter/blob/main/main.go) with tests that are failing.
When I pull out just the third line and [parse it separately](https://go.dev/play/p/C6fkNcmGxsM) it works.

I'd appreciate some advice on how to build this lexer to make it simpler to write and debug.

**Comment by alecthomas** [comment, marked as answer]
From a brief glance, one issue you're likely to have is with this type of grammar fragment:

```go
	LineOne LineOne `"Poker Hand #" @@ "\n"`
```

This is telling Participle to match the _exact_ token `Poker Hand #`, but your lexer doesn't appear to produce tokens like this. Generally the lexer should be as minimal as possible, for example in this case you might have:

```go
{"Word", `[A-Za-z]+`, nil},
{"Whitespace", `[ \t]+`, nil},
{"Newline", `\n`, nil},
{"Punct", `[#]`, nil},
```

And your grammar fragment would become something like this:

```go
	LineOne LineOne `"Poker" "Hand" "#" @@ "\n"`
```

Another issue is with the use of a stateful lexer - you almost certainly don't need this, based on the example you've given. A stateful lexer is only necessary for recursive tokens such as recursive interpolated string fragments, eg. `"hello ${coalesce(name, "Bob ${surname}")}"`. I would instead just have a simple lexer similar to the example above.

PS. The [Trace](https://pkg.go.dev/github.com/alecthomas/participle/v2#Trace) option can help when debugging the parser.

**Comment by patrobinson** [comment]
Thank you, I will give that a go.

One other problem I seem to stumble across regularly is when there is Punctuation and some value I want to extract but no word break.
For instance ` Seat 1: acd0710 ($2.34 in chips) ($2.34 in chips)`, trying to match this with a struct like:
```
type Seat struct {
	// Seat 1: acd0710 ($2.34 in chips)
	SeatNumber int    `  "Seat" @Number ":"`
	User       string `  @User`
	Stack      string `  "(" @Money "in" "chips)"`
}
```
Results in 

```
	panic: example.txt:3:17: invalid input text "($2.34 in chips)..."
```

https://go.dev/play/p/C1642UaXoGE

**Reply by alecthomas** [reply]
This means you don't have a lexer token that matches the input. I would add all your punctuation characters to  a single lexer rule, eg.

```
{"Punct", `[()$:.,]`, nil}
```

**Reply by patrobinson** [reply]
I think the problem was that I was escaping the `$` character. Thanks for your help and the great library
